src/backend/libinput.rs

```diff
@@ -0,0 +1,329 @@
+//! Implementation of input backend trait for types provided by `libinput`
+
+use backend::SeatInternal;
+use backend::input::{InputBackend, InputHandler, Seat, SeatCapabilities, MouseButton};
+use input::{Libinput, Device, Seat as LibinputSeat, DeviceCapability};
+use input::event::*;
+
+use std::io::Error as IoError;
+use std::collections::hash_map::{DefaultHasher, Entry, HashMap};
+use std::hash::{Hash, Hasher};
+
+struct SeatDesc {
+    seat: Seat,
+    pointer: (u32, u32),
+}
+
+/// Libinput based `InputBackend`.
+///
+/// Tracks input of all devices given manually or via a udev seat to a provided libinput
+/// context.
+pub struct LibinputInputBackend {
+    context: Libinput,
+    devices: Vec<Device>,
+    seats: HashMap<LibinputSeat, SeatDesc>,
+    handler: Option<Box<InputHandler<LibinputInputBackend> + 'static>>,
+    logger: ::slog::Logger,
+}
+
+impl LibinputInputBackend {
+    /// Initialize a new `LibinputInputBackend` from a given already initialized libinput
+    /// context.
+    pub fn new<L>(context: Libinput, logger: L) -> Self
+        where L: Into<Option<::slog::Logger>>
+    {
+        let log = ::slog_or_stdlog(logger).new(o!("smithay_module" => "backend_libinput"));
+        info!(log, "Initializing a libinput backend");
+        LibinputInputBackend {
+            context: context,
+            devices: Vec::new(),
+            seats: HashMap::new(),
+            handler: None,
+            logger: log,
+        }
+    }
+}
+
+impl InputBackend for LibinputInputBackend {
+    type InputConfig = [Device];
+    type EventError = IoError;
+
+    fn set_handler<H: InputHandler<Self> + 'static>(&mut self, mut handler: H) {
+        if self.handler.is_some() {
+            self.clear_handler();
+        }
+        info!(self.logger, "New input handler set.");
+        for desc in self.seats.values() {
+            trace!(self.logger, "Calling on_seat_created with {:?}", desc.seat);
+            handler.on_seat_created(&desc.seat);
+        }
+        self.handler = Some(Box::new(handler));
+    }
+
+    fn get_handler(&mut self) -> Option<&mut InputHandler<Self>> {
+        self.handler
+            .as_mut()
+            .map(|handler| handler as &mut InputHandler<Self>)
+    }
+
+    fn clear_handler(&mut self) {
+        if let Some(mut handler) = self.handler.take() {
+            for desc in self.seats.values() {
+                trace!(self.logger, "Calling on_seat_destroyed with {:?}", desc.seat);
+                handler.on_seat_destroyed(&desc.seat);
+            }
+            info!(self.logger, "Removing input handler");
+        }
+    }
+
+    fn input_config(&mut self) -> &mut Self::InputConfig {
+        &mut self.devices
+    }
+
+    fn set_cursor_position(&mut self, _x: u32, _y: u32) -> Result<(), ()> {
+        // FIXME later.
+        // This will be doable with the hardware cursor api and probably some more cases
```

Comment:
Nope. This is something provided by linux and I am not sure yet how to handle this. It will mostly likely only apply to the `DRM` backend, then again any `InputBackend` should not know anything about the underlying render implementation. (Also see http://stackoverflow.com/questions/6957039/what-is-hardware-cursor-and-how-does-it-work)

Maybe we should move `set_cursor_position` to the graphics backends?